### PR TITLE
FST-35 - adding authentication for admins

### PR DESCRIPTION
**Pull request by iprisakaru**

### CODE REVIEW

**Comment by iprisakaru** [comment]
Also fixed some bugs like making isPublic unnecessary field in creation, added DRY principles

File: src/main/scala/by/bsu/model/dao/AdminsDAO.scala [hunk 1]
```diff
@@ -2,28 +2,48 @@ package by.bsu.model.dao
...
+  def insertUniq(director: Admin): Future[Option[Admin]] = {
+    db.run(createQuery(director).asTry).map(_.toOption)
+  }
+
+  def insert(admin: Admin): Future[Admin] = {
+    db.run(admins returning admins.map(_.admin_id) += admin)
+      .map(id => admin.copy(id = Option(id)))
+  }
+
+  private def createQuery(entity: Admin): DBIOAction[Admin, NoStream, Effect.Read with Effect.Write with Effect.Transactional] = {
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/model/dao/AdminsDAO.scala:27]: This method appears in literally every DAO.
Don't Repeat Yourself

File: src/main/scala/by/bsu/model/dao/DirectorsFilmsDAO.scala [hunk 2]
```diff
@@ -30,6 +30,10 @@ class DirectorsFilmsDAO(val config: DatabaseConfig[JdbcProfile])
     }
   }
 
+  def deleteByFilmIdQuery(id: Long)={
+    directorsFilms.filter(e => e.film_id === id).delete
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/model/dao/DirectorsFilmsDAO.scala:34]: I belive there's no need to give a name to variable

File: src/main/scala/by/bsu/model/dao/FilmsDAO.scala [hunk 3]
```diff
@@ -24,7 +24,7 @@ class FilmsDAO(override val config: DatabaseConfig[JdbcProfile])
   def insertFilm(film: NewFilmWithId): Future[NewFilmWithId] = {
 
     val result = db.run(films returning films.map(_.filmId) += Film(film.id, film.name, film.ageLimit, film.shortDescription, film.timing, film.image,
-      film.releaseDate, film.awards, film.languageId, false))
+      film.releaseDate, film.awards, film.languageId, Option(false)))
       .map(id => film.copy(id = Option(id)))
     result.flatMap(data => insertLinkedTables(data)).map(data => film.copy(actorsId = data(0), genresId = data(1), directorsId = data(2), countriesId = data(3)))
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/model/dao/FilmsDAO.scala:29]: Magic numbers

File: src/main/scala/by/bsu/model/dao/FilmsDAO.scala [hunk 4]
```diff
@@ -57,10 +57,24 @@ class FilmsDAO(override val config: DatabaseConfig[JdbcProfile])
...
   def findAll(): Future[Seq[Film]] = db.run(films.filter(_.public === false).result)
 
-  def deleteById(id: Long): Future[Boolean] = {
-    db.run(films.filter(_.filmId === id).delete) map {
-      _ > 0
-    }
+
+  def deleteById(id: Long) = {
+    val e = db.run(deleteByIdLinkedTablesQuery(id))
+    e.map(_.toOption)
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/model/dao/FilmsDAO.scala:63]: Could be made a one-liner without declaring a variable

File: src/main/scala/by/bsu/model/dao/FilmsDAO.scala [hunk 5]
```diff
@@ -57,10 +57,24 @@ class FilmsDAO(override val config: DatabaseConfig[JdbcProfile])
...
+  }
+
+  private def deleteByIdLinkedTablesQuery(id: Long) = {
+    (for {
+      genresDeleted <- genresFilmsDAO.deleteByFilmIdQuery(id).asTry
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/model/dao/FilmsDAO.scala:68]: Try to set up  cascade deletes in database instead
Good that you spot a need to use transaction

File: src/main/scala/by/bsu/utils/AdminsService.scala [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import scala.concurrent.ExecutionContext.Implicits.global
+
+class AdminsService(adminsDAO: AdminsDAO) {
+
+  def getPassword(username: String) = {
+    adminsDAO.getPassword(username).map(_.password)
+  }
+
+  def create(admin: Admin) = {
+    adminsDAO.insert(admin.copy(password = admin.password.bcrypt))
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/utils/AdminsService.scala:15]: Looks good now

File: src/main/scala/by/bsu/utils/FilmsService.scala [hunk 7]
```diff
@@ -48,7 +48,7 @@ class FilmsService(filmsDAO: FilmsDAO) {
 
     val creationOfFilms = result.zip(creationOfActors.zip(creationOfGenres.zip(creationOfCountries.zip(creationOfDirectors.zip(creationOfLanguages)))))
       .flatMap(data => {
-        createWithoutFilling(NewFilmWithId(None, data._1.name, Option(data._1.ageLimit.get), data._2._1, data._2._2._1, data._2._2._2._1, data._2._2._2._2._1, data._1.shortDescription, data._1.timing, data._1.image, data._1.releaseDate, data._1.awards, data._2._2._2._2._2, false))
+        createWithoutFilling(NewFilmWithId(None, data._1.name, Option(data._1.ageLimit.get), data._2._1, data._2._2._1, data._2._2._2._1, data._2._2._2._2._1, data._1.shortDescription, data._1.timing, data._1.image, data._1.releaseDate, data._1.awards, data._2._2._2._2._2, Option(false)))
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/utils/FilmsService.scala:51]: How readable is data._2._2._2._1 from your point of view? What is the reason to use Tuple[Tuple[Tuple[Tuple[]]] datatype?

File: src/main/scala/by/bsu/web/Routes.scala [hunk 8]
```diff
@@ -2,24 +2,27 @@ package by.bsu.web
...
-      directorRoute
-    } ~ pathPrefix("actors") {
-      actorsRoute
-    } ~ pathPrefix("update-genres"){
-      updateGenres
-    } ~ pathPrefix("help-create"){
-      filmHelpRoute
+    myAuthenticateBasicAsync("none", myAuthenticator) {
+      user =>
+        LOGGER.debug(s"Admin $user makes request")
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/web/Routes.scala:17]: What request?) 
Should you log request and response bodies as well?

File: src/main/scala/by/bsu/web/api/FilmsApi.scala [hunk 9]
```diff
@@ -19,21 +20,23 @@ trait FilmJsonMapping extends DefaultJsonProtocol {
...
-        }
-        }
+        {
+          LOGGER.debug("Getting all films")
+          complete(filmsService.getAll().map(_.toJson))
+        } ~
+          (path(IntNumber)) { id => {
+            LOGGER.debug(s"Getting films with $id id")
+            complete(filmsService.getById(id).map(_.toJson))
+          }
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/web/api/FilmsApi.scala:38]: Formatting

File: src/main/scala/by/bsu/web/api/GenresApi.scala [hunk 10]
```diff
@@ -40,19 +39,23 @@ trait GenresApi extends GenresJsonMapping {
...
-      }
+    }
   }
 
   val updateGenres = {
-    post{
-      val r = genresService.getGenresFromApi.map(_.map(_.toJson))
-      complete(r)
-    }
+
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/web/api/GenresApi.scala:54]: formatting

File: src/main/scala/by/bsu/web/api/auth/HTTPBasicAuth.scala [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+import by.bsu.utils.RouteService
+import com.github.t3hnar.bcrypt._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+object HTTPBasicAuth {
+
+  def myAuthenticateBasicAsync[T](realm: String,
+                                  authenticate: (String, String) => Future[Option[T]]): Directive1[T] = {
+    def challenge = HttpChallenges.basic(realm)
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/web/api/auth/HTTPBasicAuth.scala:17]: Please make sure the standard browser dialog does not appear if I send the request via Postman with and without the Authentication header

File: src/main/scala/by/bsu/web/api/auth/HTTPBasicAuth.scala [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        onSuccess(authenticate(username, password)).flatMap {
+          case Some(client) => provide(client)
+          case None => reject(AuthenticationFailedRejection(CredentialsRejected, challenge))
+        }
+      case _ => reject(AuthenticationFailedRejection(CredentialsMissing, challenge))
+    }
+  }
+
+  def myAuthenticator(username: String, password: String): Future[Option[String]] = {
+    RouteService.adminsService.getPassword(username).map(pass=>{
```
Comment (alexander-belyaev) [inline src/main/scala/by/bsu/web/api/auth/HTTPBasicAuth.scala:29]: formatting

**Review by alexander-belyaev — CHANGES_REQUESTED** [review]

File: src/main/scala/by/bsu/utils/FilmsService.scala (hunk 7 above)
Comment (iprisakaru) [reply src/main/scala/by/bsu/utils/FilmsService.scala:51]: To make Fututre[Some] and Future[Some] to Future[(Some,Some)[

**Review by alexander-belyaev — CHANGES_REQUESTED** [review]

**Review by alexander-belyaev — APPROVED** [review]

**Pull request merged**